### PR TITLE
feat(linux): Initial fcitx5 support

### PR DESCRIPTION
**Pull request by wengxt**

### CODE REVIEW

**Comment by keyman-server** [comment]
This pull request is from an external repo and will not automatically be built. The build must still be passed before it can be merged. Ask one of the team members to make a manual build of this PR.

**Comment by wengxt** [comment]
I'm not sure if it would be better for this to live in keyman upstream or fcitx upstream, or as a separate repository. Right now keyman has its own library "keyman_kmn_processor". So for this sub project it self it doesn't need to live within the same repository.

**Comment by mcdurdin** [comment]
Thank you for your contribution! We look forward to reviewing it and working with you to integrate Keyman with fcitx5! @ermshiperete will be doing the initial code review 😁

**Review by ermshiperete — COMMENTED** [review]
@wengxt Thanks for your contribution!

I found a few issues.

I think it would be better if this would be in a separate repo under fcitx instead of the keymanapp/keyman repo.

**Comment by wengxt** [comment]
I'm totally fine to have it host under fcitx organization, but still we would need some kind of integration with km-config.

Right now there are some pieces I'm not sure about and need some opinion:
1. To put config option in dconf to share with ibus or not.
Regularly, fcitx use a ini style text config file to store data, but it can be certainly adapted to share the current dconf used by ibus for storage, though the dconf directory is "ibus".
2. Reload fcitx engine upon install new keymap. This is actually simple via a dbus call to fcitx, but still means that km-config would need some code to detect whether fcitx is being used.

**Comment by ermshiperete** [comment]
> we would need some kind of integration with km-config.

Thanks for bringing this up. It crossed my mind, but then I forgot to add it to this issue.

> 1. To put config option in dconf to share with ibus or not.

I guess it would be ok to use ini style config file when using fcitx, or maybe even in general.

> 2. Reload fcitx engine upon install new keymap. This is actually simple via a dbus call to fcitx, but still means that km-config would need some code to detect whether fcitx is being used.

How do you detect whether fcitx or ibus is used?

3. Another area where I think km-config would behave differently when fcitx is used is when installing keyboards. Currently we handle Gnome and ibus. I'm not sure how things work when fcitx is used; I would guess we might need a `install_keyboards_to_fcitx` method, and `install_keyboards_to_gnome` would also have to behave differently when running under fcitx.

**Comment by wengxt** [comment]
The detection should be fairly simple by just checking if fcitx appears on dbus. 
https://github.com/fcitx/fcitx5/blob/master/src/modules/dbus/org.fcitx.Fcitx5.service.in

I suggest not to add the engine to fcitx automatically, due to the complexity (Fcitx has two level 1: group, 2 engines in the group). I'd suggest it just simply gives a option to launch the fcitx's configuration tool and let user make the change there.

I have also considered if I just reimplement keyman-config but it seems there's already tons of code handling the extraction and installation and I don't want to duplicate that part of the code.

There's currently an additional issue with fcitx5 with current keyman-config because fcitx5 implements part of the ibus DBus API to allow application to talk to fcitx via ibus API (https://github.com/fcitx/fcitx5/tree/master/src/frontend/ibusfrontend).

 https://github.com/keymanapp/keyman/blob/340f1f72a4cbca49ddd500b543421801b60fd7e8/linux/keyman-config/keyman_config/ibus_util.py#L13 So the 10000 loop here actually would stuck for a long time (... I don't really think you need to retry here like 10000 times).

So basically I could imagine the pseudo code would be 

```
if fcitx presents:
    popup asking user to use fcitx5-configtool
elif ibus presents:
    old code here
else:
     ...
```

**Comment by ermshiperete** [comment]
I created #4910 to track this enhancement. I'll make the necessary changes in km-config as soon as I get a chance.

> There's currently an additional issue with fcitx5 with current keyman-config because fcitx5 implements part of the ibus DBus API to allow application to talk to fcitx via ibus API (https://github.com/fcitx/fcitx5/tree/master/src/frontend/ibusfrontend).
> 
> https://github.com/keymanapp/keyman/blob/340f1f72a4cbca49ddd500b543421801b60fd7e8/linux/keyman-config/keyman_config/ibus_util.py#L13
> So the 10000 loop here actually would stuck for a long time (... I don't really think you need to retry here like 10000 times).

I guess that's true :smile: I'll change that as well.

**Comment by ermshiperete** [comment]
Closing this PR since it'll be hosted under the fcitx organization.
